### PR TITLE
✨ Add bookmarks and follow api

### PR DESCRIPTION
**Pull request by williamchong**

### CODE REVIEW

File: src/routes/users/bookmarks.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
+import { Router } from 'express';
+import { userCollection as dbRef } from '../../util/firebase';
+import { filterBookmarks } from '../../util/ValidationHelper';
+import { jwtAuth } from '../../middleware/jwt';
+import { addUrlToMetadataCrawler } from '../../util/api/users/bookmarks';
+
+const uuidv4 = require('uuid/v4');
+const parse = require('url-parse');
```
Comment (nwingt) [inline src/routes/users/bookmarks.js]: `parse` is a bit confusing, how about `urlParse`?

File: src/routes/users/bookmarks.js [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+      return;
+    }
+    const query = await dbRef
+      .doc(user)
+      .collection('bookmarks')
+      .where('url', '==', url)
+      .limit(1)
+      .get();
+    if (query.docs.length) {
+      res.status(400).send('INVALID_URL');
```
Comment (nwingt) [inline src/routes/users/bookmarks.js]: Better to return status code 409, and something like `BOOKMARK_EXISTS` as error code

File: src/routes/users/bookmarks.js [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    let targetRef;
+    if (!bookmarkID) {
+      const query = await dbRef
+        .doc(user)
+        .collection('bookmarks')
+        .where('url', '==', url)
+        .limit(1)
+        .get();
+      if (!query.docs || !query.docs.length) {
+        res.status(404).send('URL_NOT_FOUND');
```
Comment (nwingt) [inline src/routes/users/bookmarks.js]: BOOKMARK_NOT_FOUND is better?

File: src/routes/users/bookmarks.js [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+      }
+      targetRef = query.docs[0].ref;
+    } else {
+      const bookmarkRef = dbRef
+        .doc(user)
+        .collection('bookmarks')
+        .doc(bookmarkID);
+      const bookmarkDoc = await bookmarkRef.get();
+      if (!bookmarkDoc.exists) {
+        res.status(404).send('URL_NOT_FOUND');
```
Comment (nwingt) [inline src/routes/users/bookmarks.js]: Same as above

File: src/routes/users/follow.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+      .set({
+        isFollowed: false,
+        ts: Date.now(),
+      }, { merge: true });
+    res.sendStatus(200);
+  } catch (err) {
+    next(err);
+  }
+});
+
```
Comment (nwingt) [inline src/routes/users/follow.js:70]: Not need to handle 404?

**Review by nwingt — CHANGES_REQUESTED** [review]

File: src/routes/users/follow.js (hunk 5 above)
Comment (williamchong) [reply src/routes/users/follow.js:70]: To mimic the behaviour of liker.land api, `DELETE` endpoint can be used to unfollow a user which was not followed in the db. This is also why I used `.set( , { merge: true )`

**Review by nwingt — APPROVED** [review]

**Pull request merged**